### ISSUE TITLE
Support "varsIgnorePattern" on no-redeclare rule

### ISSUE THREAD

**Issue by ghengeveld** [eslint/eslint#6845]
**ESLint version:** 3.2.2

**The problem**
I'm using `no-unused-vars` with `"varsIgnorePattern": "_+"` in order to allow this syntax (in React):

``` js
render() {
  const { children, beforeSubmit: _, onSuccess: __, onFailure: ___, ...props } = this.props
  return <form {...props} onSubmit={this.submit} onReset={this.reset}>{children}</form>
}
```

As you see I'm using the ES6 rest parameters syntax to create a shallow copy of the `this.props` object while omitting certain properties, so I can use the JSX spread operator to pass them all in one go. This is necessary because React will throw an error when passing non-standard attributes to an HTML element. However, some of these props are not being used in the scope of this function, so ESLint would complain about that. Right now I use the `varsIgnorePattern` option on `no-unused-vars` to allow any variable that consists of only underscores and assign each of the unused props to `_`, `__`, `___` etc. As you see this gets ugly. I would rather just alias them all to `_`, but I can't right now because `no-redeclare` would complain about that.

**Suggestion**
I suggest adding support for `"varsIgnorePattern"` to `no-redeclare`, so it can be configured like this:

``` json
"no-redeclare": [2, { "varsIgnorePattern": "_" }],
```

Background: I also program in Scala, in which it is common to use `_` do denote "something is here but I don't care what it's called" (among other uses).

I'd be happy to implement this if I know my PR has a chance of being accepted.

**Comment by eslintbot** [comment]
Hi @ghengeveld, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by platinumazure** [comment]
Can you show an example of what happens when you try to use `_` for all your variables? Does it really trigger `no-redeclare`? Please show the full ESLint output as well.

I'm surprised `_` even counts as a declaration here, to be completely honest.

**Comment by mysticatea** [comment]
Related to #6563.
Maybe @basicdays has made a rule for rest/spread properties.

**Comment by vitorbal** [comment]
@ghengeveld friendly ping, could you check out @platinumazure's question above?

**Comment by ghengeveld** [comment]
Given this code:

```js
  render() {
    const {
      children,
      onInit: _,
      onChange: _,
      beforeSubmit: _,
      onSuccess: _,
      onFailure: _,
      ...props,
    } = this.props
    return <form {...props} onSubmit={this.submit} onReset={this.reset}>{children}</form>
  }
```

The ESLint output is:
```
  93:17  error  '_' is already defined  no-redeclare
  94:21  error  '_' is already defined  no-redeclare
  95:18  error  '_' is already defined  no-redeclare
  96:18  error  '_' is already defined  no-redeclare
```

Here's the config:
```json
{
  "env": {
    "browser": true,
    "mocha": true,
    "node": true
  },
  "extends": "airbnb",
  "globals": {
    "expect": false,
    "sinon": false
  },
  "parser": "babel-eslint",
  "rules": {
    "import/no-extraneous-dependencies": ["error", {"devDependencies": true}],
    "max-len": ["error", 120],
    "no-unused-vars": ["error", { "varsIgnorePattern": "_+" }],
    "no-use-before-define": ["error", { "functions": false }],
    "react/jsx-filename-extension": ["error", { "extensions": [".js"] }],
    "semi": ["error", "never"]
  }
}
```

**Comment by platinumazure** [comment]
I still am having a hard time understanding how `_` counts as a declaration in this code. I feel like we have a fundamental bug in no-redeclare.

@michaelficarra Can you help me understand? In the code in [this comment](https://github.com/eslint/eslint/issues/6845#issuecomment-242414794), should those object pattern values (`_`) count as variable declarations in the first place, since they are values not keys?

**Comment by ghengeveld** [comment]
I think you're confusing my code for a regular object literal, while what I'm doing is a destructuring assignment, which is in fact a variable declaration, so no-redeclare is correct. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names

**Comment by ghengeveld** [comment]
As I was trying to provide an example of what the destructuring would be compiled to by Babel, I found that Babel will in fact not compile this code:

```
TypeError: /Users/ghengeveld/Development/Prive/Libs/react-isomorphic-form/src/Form.js: Duplicate declaration "_"
  91 |       children,
  92 |       onInit: _,
> 93 |       onChange: _,
     |                 ^
  94 |       beforeSubmit: _,
  95 |       onSuccess: _,
  96 |       onFailure: _,
```

I can understand why Babel would disallow re-assigning a const, but I found that it also doesn't allow re-assingment when I use a `let`. However, `var` does work. Try this in [Babel's REPL](https://babeljs.io/repl/#?evaluate=true&lineWrap=false&presets=es2015%2Cstage-2&experimental=true&loose=false&spec=false&playground=false&code=var%20%7B%0A%20%20children%2C%0A%20%20onInit%3A%20_%2C%0A%20%20onChange%3A%20_%2C%0A%20%20beforeSubmit%3A%20_%2C%0A%20%20onSuccess%3A%20_%2C%0A%20%20onFailure%3A%20_%2C%0A%7D%20%3D%20obj.props%0A):

```
var {
  children,
  onInit: _,
  onChange: _,
  beforeSubmit: _,
  onSuccess: _,
  onFailure: _,
} = obj.props
```

**Comment by platinumazure** [comment]
Ah, yes, you're correct. For some reason I thought it was the other way
around (variable name first, then object property name). Okay, I agree this
isn't a bug but an enhancement request. I could get behind the enhancement.

**Comment by ghengeveld** [comment]
Because I'm also using the `no-var` rule, ESLint will error on the example provided in my previous comment. However, since neither `let` nor `const` is valid in this scenario, `no-var` should probably not be complaining in this particular situation...

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.

**Comment by sleepycat** [comment]
@kaicataldo This should be reopened IMO. @ghengeveld is making an useful enhancement request and even offering to implement, and @platinumazure seems to support this. I personally would love to see this option added.
